frontends/krita/krita_diff/utils.py

```diff
@@ -232,9 +241,39 @@ def reset_docker_layout():
         qmainwindow.addDockWidget(Qt.LeftDockWidgetArea, d)
 
     qmainwindow.tabifyDockWidget(dockers[TAB_SDCOMMON], dockers[TAB_CONFIG])
+    qmainwindow.tabifyDockWidget(dockers[TAB_SDCOMMON], dockers[TAB_CONTROLNET])
     qmainwindow.tabifyDockWidget(dockers[TAB_SDCOMMON], dockers[TAB_PREVIEW])
     qmainwindow.tabifyDockWidget(dockers[TAB_TXT2IMG], dockers[TAB_IMG2IMG])
     qmainwindow.tabifyDockWidget(dockers[TAB_TXT2IMG], dockers[TAB_INPAINT])
     qmainwindow.tabifyDockWidget(dockers[TAB_TXT2IMG], dockers[TAB_UPSCALE])
     dockers[TAB_SDCOMMON].raise_()
     dockers[TAB_INPAINT].raise_()
+
+def remove_unmasked_content_for_inpaint(img, mask):
```

Comment:
This is very slow for large images, better approach is to insert mask as transparency layer: https://api.kde.org/krita/html/classTransparencyMask.html.

Reply:
Currently working on this, but can't find a way to import this class to the script. Doesn't seem to work from krita module. No error is shown but it messes up the whole plugin.

Reply:
Hello. It seems Google Colab has changed their ToS to prohibit the use of remote UIs. This means I'm unable to perform testing and use the plugin, so unfortunately I'll have to stop development in this implementation. Anyone interested can rescue it and follow development.

Reply:
that's so unfortunate :(

is it possible for you to use Runpod instead of colab @JasonS09 ? it's generally affordable to use, and i can even sponsor your GPU hours if you want

Reply:
Hey, thank you for the offer. Honestly, I don't want to pay for this since I'm broke (got 17 dollars in my bank account, no incomes). If you say you can sponsor that could be an option, but I'm not sure if that service provider allows remote connections to the UI. Last time I tried in Paperspace, I couldn't use it as backend (simply didn't work even with the --api flag), and after a while I got kicked from my session, then unable to login again. I'm suspect they banned me for trying to use the UI as backend.

Reply:
Glad to hear! But yea i have no idea about backend and frontend so half of what you said didn't make sense to me 😅  just let me know if you want to give it a try still. Also maybe add --medvram and --xformers flag so you can run it faster on a less powerful system?

Reply:
I might try and pick it up next month when I'm done with internship. But I might also decide to work on LLMs instead.

Reply:
Hello. I have been trying to implement the transparency layer approach for the inpaint mask but so far I have failed to do so. Mostly because Krita API seems to have an issue when setting active layers and then running actions (to create the mask and then flatten the layer). I'm running out of options.

Reply:
> Hello. I have been trying to implement the transparency layer approach for the inpaint mask but so far I have failed to do so. Mostly because Krita API seems to have an issue when setting active layers and then running actions (to create the mask and then flatten the layer). Workarounds are very rigid and doubt they're practical. I'm running out of options.

I know it has been a while, but how have you been attempting to get the mask?  I've probed around in the API a bit and found that you can get the API to return the mask used if you add `"return_mask": true` to the override settings.  Every other image returned in the payload will be the inpainting mask used for the image.  Adding a layer as a transparency mask layer from the payload should be as simple of a matter as adding the image layers themselves, shouldn't it?  The only side effect is that these masks will show up in the image grid (if that is used).

Reply:
> I know it has been a while, but how have you been attempting to get the mask? I've probed around in the API a bit and found that you can get the API to return the mask used if you add `"return_mask": true` to the override settings. Every other image returned in the payload will be the inpainting mask used for the image. Adding a layer as a transparency mask layer from the payload should be as simple of a matter as adding the image layers themselves, shouldn't it? The only side effect is that these masks will show up in the image grid (if that is used).
> 
> EDIT: I am attempting to implement this myself but I have never worked with Krita's API before. I am pretty sure there's nothing stopping me from retrofitting the transparency mask code that already exists to do this, but any advice or guidance would be appreciated.

The problem I am facing is the following:

- Creating a transparency mask by `self.doc.createNode(name, "transparencyMask")`  in `script.py` is not working properly. It creates a node but you can't do anything with it (not really a transparency mask).
-  `self.doc.createTransparencyMask(name)` doesn't seem to be a thing, even though you can [find it in the documentation](https://api.kde.org/krita/html/classDocument.html#abbd8e5ca62dd2952623c2d5cbc6faf5f)
- Alternatively, it is possible to create one by calling actions `self.app.action("add_new_transparency_mask")`. However, you seemingly can' use `setPixelData()` [to draw the mask into the mask layer](https://api.kde.org/krita/html/classNode.html#a4e0b624db748aa8cf63ba84131dfc1a7). Or at least all my attempts have failed to do so.
- So this only leaves me with one option I can think of: create a paint layer first, set the mask pixel data, then convert it into a transparency mask. However, there is another issue with this approach. Setting active node with `self.doc.setActiveNode(layer)` will work but not for the actions like `self.app.action("convert_to_transparency_mask")`. They will completelly ignore the current active node even if it's explicitly set. Making it really difficult to work with.

Reply:
>     * Creating a transparency mask with `self.doc.createNode(name, "transparencyMask")`  in `script.py` is not working properly. It creates a node but you can't do anything with it (not really a transparency mask).
> 
>     * `self.doc.createTransparencyMask(name)` doesn't seem to be a thing, even though you can [find it in the documentation](https://api.kde.org/krita/html/classDocument.html#abbd8e5ca62dd2952623c2d5cbc6faf5f).
> 
>     * Alternatively, it is possible to create one by calling actions `self.app.action("add_new_transparency_mask")`. However, you seemingly can' use `setPixelData()` [to draw the mask into the mask layer](https://api.kde.org/krita/html/classNode.html#a4e0b624db748aa8cf63ba84131dfc1a7). Or at least all my attempts have failed to do so.
> 
>     * So this only leaves me with one option I can think of: create a paint layer first, set the mask pixel data, then convert it into a transparency mask. However, there is another issue with this approach. Setting active node with `self.doc.setActiveNode(layer)` will work but not for the actions like `self.app.action("convert_to_transparency_mask")`. They will completelly ignore the current active node even if it's explicitly set. Making it really difficult to work with.

Well... I've independently verified most of these.

On my most successful attempt so far, where I attempted to retrofit `transparency_mask_inserter(self)` to handle these masks, I managed to get four functioning transparency masks on a batch and wrote data to one of them, though it did not appear correct, possibly because it needs to be converted to grayscale first?  I did that by passing the mask all the way in that function and putting it right after `add_mask_action.trigger()` which is far less than ideal...  I suspect that the issues you're encountering with setting pixel data on a layer created through app.action are more related to the race conditions that the inserter function is meant to work around, though, based on this experience.

I'm going to see if I can set pixel data on masks without completely butchering the mask inserter function, I'll let you know if I find anything out.

Reply:
> > I've made a pull request on your branch for a partially working implementation...
> 
> Can you point me to it? I can't find it.

I seem to have opened the pull request on my own fork... Should be fixed now.

Reply:
I figured it out.  Fully working implementation.

setPixelData on a node is broken for some unexplainable reason.  But you can setPixelData on a Selection, and then call to create a transparency mask... because the selection will become the content of that mask.  This also seems to handily make every race condition irrelevant.

I am going to clean it up a bit, and then I will PR what should be a complete implementation